keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1244 @@
+
+### Plan
+
+Because `generic.StoreOptions` consumes selectors from built-in and CRDs in the
+same way, we would extend the `GetAttrs` of CRDs to offer fields for selection.
+https://github.com/jpbetz/kubernetes/tree/crd-object-filters contains a working
+prototype.
+
+We would leverage the fieldPath validation supported added for CRD Validation
+Rules to validate paths.
+
+We would leverage the jsonpath library in the same way as used by
+additionalPrinterColumns to retrieve values from custom resources using
+jsonpaths.
+
+The majority of the remaining work will be to update the API to add the
+field and to validate it.
+
+#### API
+
+```go
+type CustomResourceDefinitionVersion struct {
+  
+  // selectableFields specifies paths to fields that may be used as field selectors.
+  // A maximum of 8 selectable fields are allowed.
+	// See https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors
+	SelectableFields []SelectableField `json:"selectableFields,omitempty" protobuf:"bytes,9,rep,name=selectableFields"`
+}
+
+  // SelectableField specifies the JSON path of a field that may be used with field selectors.
+type SelectableField struct {
+	// jsonPath is a simple JSON path which is evaluated against each custom resource to produce a
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a valid field of type string, boolean or integer. Strings with enum values
+  // and strings with formats are allowed.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+
+  // default provides a value for field selection if the value referred to
+  // by jsonPath is absent.
+  // default is required unless all parts of jsonPath are required fields, or have defaults.
+  // +Optional
+  Default *string `json:"default" protobuf:"bytes,2,opt,name=default"`
+}
+```
+
+##### Validaiton rules
+
+- `selectableFields[].jsonPath` must be a "simple path" (similar to
+  `additionalPrinterColumns[].jsonPath` except validated properly). I.e.
+    property names separated by '.', e.g. `.spec.x.y` (The path may not contain
+    map or array indexing)
+- The schema type referenced by `selectableFields[].jsonPath` must be one of:
+  string, integer, boolean (enums and fields with string formats are allowed)
+- If (and only if) the schema type referenced by `selectableFields[].jsonPath`
+  is optional (directly or indirectly), a `selectableFields[].ifAbsent` field
+  must provide a fallback string value.
+- A maximum of 8 `selectableFields` per CRD version are allowed.
+- `selectableFields` may not refer to metadata fields.
+- `selectableFields` may not contain duplicates.
+
+
+#### Discovery
+
+We will add discovery details in to OpenAPIv3:
+
+```json
+"org.example.v1.CustomResource": {
+        "type": "object",
+        "x-kubernetes-group-version-kind": [ ... ],
+        "x-kubernetes-selectable-fields": [
```

Comment:
@Jefftree and @sttts for openapi extension details.

Reply:
lgtm on openapi extension naming, added one question comment